### PR TITLE
[FLINK-31795] Add config to disable savepoint cleanup

### PR DESCRIPTION
**Pull request by darenwkt**
## What is the purpose of the change

Add new operator config "savepoint.cleaner.enabled" to allow user to configure whether to enable SavepointObserver to clean up savepoint history.

## Brief change log
- Added new operator config.

## Verifying this change

Added unit test to verify savepoint history is not cleaned up when config is set.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: (no)
  - Core observer or reconciler logic that is regularly executed: (yes) - SavepointObserver logic

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java [hunk 1]
```diff
@@ -59,10 +59,11 @@ public class FlinkOperatorConfiguration {
     Duration flinkCancelJobTimeout;
     Duration flinkShutdownClusterTimeout;
     String artifactsBaseDir;
+    boolean savepointCleanerEnabled;
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:62]: I think this should not be part of the operator configuration (which are system wide configs) but instead should be configurable on a per resource level (simply get it based on the observeconfig)

**Review by gyfora — CHANGES_REQUESTED** [review]

**Comment by gyfora** [comment]
@darenwkt I think it would make sense to finish up this PR, this can be a valuable feature for some users if they want to do manual savepoint cleanup.

File: docs/layouts/shortcodes/generated/dynamic_section.html [hunk 2]
```diff
@@ -98,6 +98,12 @@
             <td>Boolean</td>
             <td>Configure the array merge behaviour during pod merging. Arrays can be either merged by position or name matching.</td>
         </tr>
+        <tr>
+            <td><h5>kubernetes.operator.savepoint.cleaner.enabled</h5></td>
```
[the remark this answers is not in the document]
Comment (darenwkt) [reply docs/layouts/shortcodes/generated/dynamic_section.html:102]: Renamed to `cleanup` in latest commit

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java (hunk 1 above)
Comment (darenwkt) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:62]: Move to ObserverConfig

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SavepointObserver.java [hunk 3]
```diff
@@ -142,6 +142,11 @@ void cleanupSavepointHistory(
             SavepointInfo currentSavepointInfo,
             Configuration observeConfig) {
 
+        if (!configManager.getOperatorConfiguration().isSavepointCleanerEnabled()) {
+            LOG.info("Operator savepoint cleaner is not enabled, do not clean savepointHistory");
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (darenwkt) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SavepointObserver.java:151]: Agreed, I have updated the logic to remove from savepoint history/status even if cleanup is disabled. This ensures savepointHistory in CR is within maxCount and maxAge. The only difference is cleanup disabled will not send `disposeSavepoint` REST call to JM.

**Comment by darenwkt** [comment]
@gyfora , thank you for reviewing the PR, I have addressed the comments and changes in the latest commit, thank you

**Comment by gyfora** [comment]
Looks great! Thank you @darenwkt :)

**Pull request merged**